### ISSUE TITLE
[InstCombine] reduce extractelt of select-of-constants

### ISSUE THREAD

**Issue by rotateright** [llvm/llvm-project#58763]
```ll
define i8 @src(i1 %b) {
  %s = select i1 %b, <3 x i8> <i8 -1, i8 34, i8 23>, <3 x i8> <i8 1, i8 2, i8 4>
  %e = extractelement <3 x i8> %s, i64 2
  ret i8 %e
}

define i8 @tgt(i1 %b) {
  %r = select i1 %b, i8 23, i8 4
  ret i8 %r
}
```

https://alive2.llvm.org/ce/z/Nx7Ffg

This pattern appears (via bogus SLP transforms?) in the example from issue #58552.

**Comment by tsymalla** [comment]
Working on this.

**Comment by tsymalla** [comment]
https://reviews.llvm.org/B197536

**Comment by tsymalla** [comment]
After that patch is being applied, e. g. after `visitExtractElementInst` calls `FoldOpIntoSelect`, constant folding of select operands will fail in sequences like that:
```
%oob = select i1 %c, i32 3, i32 4
%s = select i1 %c, <3 x i32> <i32 2, i32 3, i32 4>, <3 x i32> <i32 5, i32 6, i32 7>
%r = extractelement <3 x i32> %s, i32 %oob
```

This is part of another issue that needs to be treated separately.

**Comment by nikic** [comment]
Patch has landed in 470aea5ed403d43b68e2ffdbec451d79913eb15d.
